hw lesson 17 complete

mctapok opened the pull request.

mctapok committed:
hw lesson 17 complete

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    public E getLast() {
+        System.out.println(tail.data);
```
Сейчас придираться не буду но лишнюю отладочную печать надо убирать.

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+        return head.data;
+    }
+
+    public void show() {
```
Вообще таких методов не должно быть в структурах данных. Если мы хотим что-то распечатать, то нужно внешними относительно класса средствами пользоваться

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+    public void add(int position, E data) {
+        Node<E> newNode = new Node<>(data);
+        if (position == 0) {
+            newNode.nextNode = head;
```
а хеду предыдущую не нужно дать тут ?

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+
+        newNode.nextNode = currentNode.nextNode;
+        currentNode.nextNode = newNode;
+    }
```
пустая строка

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+    public E remove(int position) {
+        Node<E> currentNode = head;
+        if (position == 0) {
+            head = head.nextNode;
```
если список пустой и head null, Вы тут нпе получите, а хотелось бы более осмысленную ошибку

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+        if (currentNode == null) {
+            throw new IndexOutOfBoundsException("Позиция больше размера списка");
+        }
+        currentNode.nextNode = currentNode.nextNode.nextNode;
```
а если это последний элемент - currentNode != null, currentNode.nextNode == null

Вообще я бы дебагом проверил, все ли ссылки у Вас выставляются и более внимательно протестировал работу функций

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyLinkedList.java:
```diff
@@ -0,0 +1,127 @@
+        return currentNode.data;
+    }
+
+    public int getSize() {
```
вообще size лучше просто менять каждый раз при добавлении и удалении а не бегать по нодам.

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyQueue.java:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    public void poll(){
+        linkedList.remove(linkedList.getSize());
```
Тут у меня вопрос с getSize - если мы с нуля нумеруем, она всегда будет на единицу выше индекса последгнего элемента

idontgiveit replied:
Если мы не можем получить элемент из очереди - очередь бесполезна. Элемент нужно возвращать

idontgiveit commented on src/main/java/ru/gavrilov/basic/homework/lesson17/MyStack.java:
```diff
@@ -0,0 +1,15 @@
+    public void push(E data){
+        myLinkedList.addFirst(data);
+    }
+    public void pop(){
```
То же что и для очереди - элемент нужно возвращать иначе в стеке нет никакого смысла.

mctapok committed:
hw lesson28 complete

mctapok committed:
lesson 17
изменен метод remove;
изменен метод add;
добавлено поле size;
изменен метод getsize;

mctapok commented:
перелопатил метод remove, откорректировал остальные методы и присваивание ссылок
